File: frontend/src/components/today/todayRitualSignals.ts
/**
 * Сигналы ритуала «Сегодня»: possible / avoid / support и разбор day_layer.
 * Паритет с `TodayRitualSignals.swift` (iOS).
 */
import {
  isGarbageRitualActionCue,
  replaceQuotedEnSlugsForRuDisplay,
  repairRitualDoNotEnterLine,
} from "@/components/today/ritualCueSanitizer";

export const RITUAL_GUIDANCE_SUMMARY_FALLBACK =
  "Открой день одним действием, которое убирает шум.";

function trimS(s: string | null | undefined): string {
  return (s || "").trim();
}

export type MorningLike = {
  daily_recommendations?: {
    what_to_do?: string | null;
    what_to_avoid?: string | null;
    priorities?: unknown;
  } | null;
  daily_horoscope?: {
    spine?: {
      best_mode?: string | null;
      first_move?: string | null;
      do_not_enter?: string | null;
    } | null;
  } | null;
  daily_forecast_summary?: { summary?: string | null } | null;
} | null | undefined;

export function firstPriorityLine(morning: MorningLike): string | null {
  const raw = morning?.daily_recommendations?.priorities;
  if (!Array.isArray(raw)) return null;
  for (const x of raw) {
    if (typeof x === "string") {
      const t = x.trim();
      if (t) return t;
    }
  }
  return null;
}

/** Как в iOS: what_to_do → first_move → forecast (до 3 строк, без дублей). */
export function buildRitualPossibleSignals(morning: MorningLike): string[] {
  const a: string[] = [];
  const wtd = replaceQuotedEnSlugsForRuDisplay(trimS(morning?.daily_recommendations?.what_to_do));
  if (wtd && !isGarbageRitualActionCue(wtd)) a.push(wtd);
  const fm = replaceQuotedEnSlugsForRuDisplay(trimS(morning?.daily_horoscope?.spine?.first_move));
  if (fm && !isGarbageRitualActionCue(fm) && !a.includes(fm)) a.push(fm);
  const summary = replaceQuotedEnSlugsForRuDisplay(trimS(morning?.daily_forecast_summary?.summary));
  if (summary && !isGarbageRitualActionCue(summary) && a.length < 3) a.push(summary);
  return a.slice(0, 3);
}

export function buildRitualAvoidSignals(morning: MorningLike): string[] {
  const a: string[] = [];
  const wta = replaceQuotedEnSlugsForRuDisplay(trimS(morning?.daily_recommendations?.what_to_avoid));
  if (wta) a.push(wta);
  const dneRaw = trimS(morning?.daily_horoscope?.spine?.do_not_enter);
  const dne = dneRaw ? repairRitualDoNotEnterLine(dneRaw) : "";
  if (dne && !a.includes(dne)) a.push(dne);
  return a;
}

/**
 * Если есть best_mode — три опоры; иначе одна строка (как `dashboard.guidanceSummary` на iOS).
 */
export function buildRitualSupportSignals(bestMode: string, guidanceSummary: string): string[] {
  const b = trimS(bestMode);
  if (b) return [b, "Тишина", "Структура"];
  const g = trimS(guidanceSummary);
  return [g || RITUAL_GUIDANCE_SUMMARY_FALLBACK];
}

export function guidanceSummaryForRitual(args: {
  morning: MorningLike;
  actionPlanFirstLine: string | null | undefined;
}): string {
  const p = firstPriorityLine(args.morning);
  if (p) return p;
  const first = trimS(args.actionPlanFirstLine);
  if (first) return first;
  return RITUAL_GUIDANCE_SUMMARY_FALLBACK;
}

function str(v: unknown): string | null {
  if (typeof v !== "string") return null;
  const t = v.trim();
  return t || null;
}

export type ParsedDayLayerNarrative = {
  nudgeMessage: string | null;
  personalTitle: string | null;
  personalBody: string | null;
  chips: string[];
  lifeWeekly: string | null;
  lifeDiscipline: string | null;
  questionPrompt: string | null;
  miniDecisionCaption: string | null;
};

/** Поля payload поверхности `day_layer` из `/today/narrative`. */
export function parseDayLayerPayload(
  payload: Record<string, unknown> | null | undefined,
): ParsedDayLayerNarrative {
  if (!payload) {
    return {
      nudgeMessage: null,
      personalTitle: null,
      personalBody: null,
      chips: [],
      lifeWeekly: null,
      lifeDiscipline: null,
      questionPrompt: null,
      miniDecisionCaption: null,
    };
  }
  const chipsRaw = payload.personal_insight_chips;
  const chips = Array.isArray(chipsRaw)
    ? chipsRaw.map((x) => (typeof x === "string" ? x.trim() : "")).filter(Boolean)
    : [];
  return {
    nudgeMessage: str(payload.nudge_message),
    personalTitle: str(payload.personal_insight_title),
    personalBody: str(payload.personal_insight_body),
    chips,
    lifeWeekly: str(payload.life_now_weekly),
    lifeDiscipline: str(payload.life_now_discipline),
    questionPrompt: str(payload.question_of_day_prompt),
    miniDecisionCaption: str(payload.mini_decision_caption),
  };
}

export function dayLayerHasContent(p: ParsedDayLayerNarrative): boolean {
  return !!(
    p.nudgeMessage ||
    p.personalTitle ||
    p.personalBody ||
    p.chips.length ||
    p.lifeWeekly ||
    p.lifeDiscipline ||
    p.questionPrompt ||
    p.miniDecisionCaption
  );
}

File: ios/TodayFlow/TodayFlow/Design/TodayRitualSignals.swift
import Foundation

/// Сигналы «возможно / избегать / опоры» для ритуала Today. Паритет с `frontend/src/components/today/todayRitualSignals.ts`.
enum TodayRitualSignals {
    static let guidanceFallback = "Открой день одним действием, которое убирает шум."

    /// Паритет с `buildDayEnergySummary` в `todayPageUtils.ts` (поле `guidance`).
    static func dayEnergyGuidanceLine(score: Int) -> String {
        let s = min(100, max(0, score))
        if s <= 30 {
            return "Сбавь темп и двигайся короткими шагами."
        }
        if s <= 70 {
            return "Держи ровный темп и закрывай задачи по одной."
        }
        return "Ресурса больше обычного. Держи фокус на главном."
    }

    /// Паритет с `guideHeadline` на вебе: первый сегмент до `[.!?]\\s+`, иначе префикс 140 символов.
    static func guideHeadlineSegment(fromForecastSummary summary: String) -> String {
        let trimmed = summary.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !trimmed.isEmpty else { return "" }
        guard let regex = try? NSRegularExpression(pattern: "[.!?]\\s+", options: []) else {
            return String(trimmed.prefix(140))
        }
        let range = NSRange(trimmed.startIndex ..< trimmed.endIndex, in: trimmed)
        guard let match = regex.firstMatch(in: trimmed, options: [], range: range), match.range.location > 0 else {
            return String(trimmed.prefix(140))
        }
        let head = (trimmed as NSString).substring(with: NSRange(location: 0, length: match.range.location))
        let h = head.trimmingCharacters(in: .whitespacesAndNewlines)
        return h.isEmpty ? String(trimmed.prefix(140)) : h
    }

    static func possible(from morning: TodayMorning?) -> [String] {
        var a: [String] = []
        let wtd = TodayRitualCueSanitizer.replaceQuotedEnSlugsForRuDisplay(
            morning?.dailyRecommendations?.whatToDo?.trimmingCharacters(in: .whitespacesAndNewlines) ?? ""
        )
        if !wtd.isEmpty, !TodayRitualCueSanitizer.isGarbageRitualActionCue(wtd) { a.append(wtd) }
        let fm = TodayRitualCueSanitizer.replaceQuotedEnSlugsForRuDisplay(
            morning?.dailyHoroscope?.spine?.firstMove?.trimmingCharacters(in: .whitespacesAndNewlines) ?? ""
        )
        if !fm.isEmpty, !TodayRitualCueSanitizer.isGarbageRitualActionCue(fm), !a.contains(fm) { a.append(fm) }
        let summary = TodayRitualCueSanitizer.replaceQuotedEnSlugsForRuDisplay(
            morning?.dailyForecastSummary?.summary?.trimmingCharacters(in: .whitespacesAndNewlines) ?? ""
        )
        if !summary.isEmpty, !TodayRitualCueSanitizer.isGarbageRitualActionCue(summary), a.count < 3 { a.append(summary) }
        return Array(a.prefix(3))
    }

    static func avoid(from morning: TodayMorning?) -> [String] {
        var a: [String] = []
        let wta = TodayRitualCueSanitizer.replaceQuotedEnSlugsForRuDisplay(
            morning?.dailyRecommendations?.whatToAvoid?.trimmingCharacters(in: .whitespacesAndNewlines) ?? ""
        )
        if !wta.isEmpty { a.append(wta) }
        let dneRaw = morning?.dailyHoroscope?.spine?.doNotEnter?.trimmingCharacters(in: .whitespacesAndNewlines) ?? ""
        let dne = TodayRitualCueSanitizer.repairRitualDoNotEnterLine(dneRaw)
        if !dne.isEmpty, !a.contains(dne) { a.append(dne) }
        return a
    }

    /// При непустом `bestMode` — три строки; иначе одна опора (`guidanceSummary` с дашборда, как на вебе).
    static func support(bestMode: String, guidanceSummary: String) -> [String] {
        let b = bestMode.trimmingCharacters(in: .whitespacesAndNewlines)
        if !b.isEmpty { return [b, "Тишина", "Структура"] }
        let g = guidanceSummary.trimmingCharacters(in: .whitespacesAndNewlines)
        return [g.isEmpty ? Self.guidanceFallback : g]
    }
}

// MARK: - «Почему так?» — строки как на вебе (`TodayRitualFlow.tsx` → `buildWhyReasonLines`)

enum TodayRitualWhyReasonLines {
    /// Тот же порядок кандидатов и фильтры дубликатов, что в `frontend/.../TodayRitualFlow.tsx`.
    static func build(
        summaryTitle: String,
        possible: [String],
        guidePayload: [String: JSONValue]?,
        spineMainRisk: String?,
        spineBestMode: String?,
        numerologyDisplayValue: String,
        numerologyReduced: Int?,
        numerologyDetailLine: String?,
        morningFocus: String?,
        whyMoon: String?,
        whyLunar: String?,
        avoidLines: [String]
    ) -> (headline: String?, lines: [String]) {
        let rawHeadline = narrativeString(guidePayload, "headline") ?? narrativeString(guidePayload, "focus_line")
        let headline: String?
        if let h = rawHeadline, !TodayPersonalDayRhythm.lineRedundantWithAny(h, pool: avoidLines) {
            headline = h
        } else {
            headline = nil
        }

        var candidates: [String] = []

        let phaseLabel = TodayPersonalDayRhythm.normalizeLunarPhaseLabel(whyMoon)
        let themes = whyLunar?.trimmingCharacters(in: .whitespacesAndNewlines) ?? ""
        var moonBlock = ""
        if !themes.isEmpty, let p = phaseLabel, !p.isEmpty {
            moonBlock = "\(themes) · \(p)"
        } else if !themes.isEmpty {
            moonBlock = themes
        } else if let p = phaseLabel {
            moonBlock = p
        }

        let lunarInf = narrativeString(guidePayload, "lunar_influence")
        if let lunarInf, !moonBlock.isEmpty, TodayPersonalDayRhythm.lunarInfluenceRedundantWithAstronomy(lunarInf, moonBlock: moonBlock) {
            candidates.append(clipWhyLine(moonBlock, maxLen: 420))
        } else if let lunarInf, !moonBlock.isEmpty {
            candidates.append(clipWhyLine(moonBlock, maxLen: 420))
            if !TodayPersonalDayRhythm.lunarInfluenceRedundantWithAstronomy(lunarInf, moonBlock: moonBlock) {
                candidates.append(clipWhyLine(lunarInf, maxLen: 420))
            }
        } else if let lunarInf {
            candidates.append(clipWhyLine(lunarInf, maxLen: 420))
        } else if !moonBlock.isEmpty {
            candidates.append(clipWhyLine(moonBlock, maxLen: 420))
        }

        if let risk = spineMainRisk?.trimmingCharacters(in: .whitespacesAndNewlines), !risk.isEmpty {
            candidates.append("Напряжение дня сейчас в одном месте: \(risk)")
        }
        if let best = spineBestMode?.trimmingCharacters(in: .whitespacesAndNewlines), !best.isEmpty {
            candidates.append("Лучше всего день раскрывается через такой режим: \(best)")
        }
        if let transit = narrativeString(guidePayload, "transit_influence") {
            candidates.append(clipWhyLine(transit, maxLen: 420))
        }
        let detail = numerologyDetailLine?.trimmingCharacters(in: .whitespacesAndNewlines) ?? ""
        if detail.count >= 36 {
            candidates.append(clipWhyLine(detail, maxLen: 420))
        } else {
            candidates.append(TodayPersonalDayRhythm.personalDayRhythmLine(reduced: numerologyReduced, displayValue: numerologyDisplayValue))
        }
        if let nextAction = narrativeString(guidePayload, "next_action") {
            candidates.append(nextAction)
        }
        if let focus = morningFocus?.trimmingCharacters(in: .whitespacesAndNewlines), !focus.isEmpty,
           !TodayPersonalDayRhythm.isDiscardableMorningFocus(focus) {
            candidates.append("Утренний фокус для дня: \(focus)")
        }
        if let nudge = narrativeString(guidePayload, "nudge_message") {
            candidates.append(nudge)
        }

        var seenNorms = Set<String>()
        var filtered: [String] = []
        let avoidPool = avoidLines.map { $0.trimmingCharacters(in: .whitespacesAndNewlines) }.filter { !$0.isEmpty }
        for line in candidates {
            let t = line.trimmingCharacters(in: .whitespacesAndNewlines)
            guard !t.isEmpty else { continue }
            guard !isDuplicateWhyLine(line: t, summaryTitle: summaryTitle, possible: possible) else { continue }
            let norm = normalizeWhy(t)
            guard !seenNorms.contains(norm) else { continue }
            guard !TodayPersonalDayRhythm.lineRedundantWithAny(t, pool: avoidPool + filtered) else { continue }
            seenNorms.insert(norm)
            filtered.append(t)
        }

        var outHeadline = headline
        if let h = outHeadline, TodayPersonalDayRhythm.lineRedundantWithAny(h, pool: filtered) {
            outHeadline = nil
        }
        return (outHeadline, Array(filtered.prefix(5)))
    }

    private static func clipWhyLine(_ s: String, maxLen: Int) -> String {
        let t = s.trimmingCharacters(in: .whitespacesAndNewlines)
        if t.count <= maxLen { return t }
        let idx = t.index(t.startIndex, offsetBy: max(0, maxLen - 1))
        return String(t[..<idx]) + "…"
    }

    private static func narrativeString(_ payload: [String: JSONValue]?, _ key: String) -> String? {
        guard let raw = payload?[key]?.stringValue else { return nil }
        let s = raw.trimmingCharacters(in: .whitespacesAndNewlines)
        return s.isEmpty ? nil : s
    }

    private static func normalizeWhy(_ s: String) -> String {
        let trimmed = s.trimmingCharacters(in: .whitespacesAndNewlines).lowercased()
        guard let regex = try? NSRegularExpression(pattern: "\\s+", options: []) else {
            return trimmed
        }
        let range = NSRange(trimmed.startIndex ..< trimmed.endIndex, in: trimmed)
        return regex.stringByReplacingMatches(in: trimmed, options: [], range: range, withTemplate: " ")
    }

    private static func isDuplicateWhyLine(line: String, summaryTitle: String, possible: [String]) -> Bool {
        let n = normalizeWhy(line)
        guard n.count >= 14 else { return false }
        let sum = normalizeWhy(summaryTitle)
        if sum.count > 10 {
            let sumPrefixLen = min(52, sum.count)
            let nPrefixLen = min(52, n.count)
            let sumPrefix = String(sum.prefix(sumPrefixLen))
            let nPrefix = String(n.prefix(nPrefixLen))
            if n == sum || n.contains(sumPrefix) || sum.contains(nPrefix) {
                return true
            }
        }
        for p in possible {
            let pn = normalizeWhy(p)
            guard pn.count > 12 else { continue }
            if n == pn { return true }
            if pn.count > 20, n.contains(pn) { return true }
            if n.count > 20, pn.contains(n) { return true }
        }
        return false
    }
}
